Libraries/pbxbuild/Sources/Phase/ProductTypeResolver.cpp

```diff
@@ -268,6 +269,135 @@ ResolveFrameworkStructure(Phase::Environment const &phaseEnvironment, Phase::Con
+             */
+            auto contents = std::string();
+            contents += "framework module " + moduleName + " {\n";
+            contents += "  umbrella header \"" + *umbrellaHeaderName + "\"\n";
+            contents += "  \n";
+            contents += "  export *\n";
+            contents += "  module * { export * }\n";
+            contents += "}\n";
+
+            moduleMapAuxiliaryFile = Tool::Invocation::AuxiliaryFile(
+                intermediateModuleMapPath,
+                std::vector<uint8_t>(contents.begin(), contents.end()),
+                false);
+        } else {
+            fprintf(stderr, "warning: target defines module, but has no umbrella header\n");
+        }
+    }
+
+    if (moduleMapAuxiliaryFile) {
+        /* Define source module map. */
+        // TODO: it would be nicer to attach this to the copy invocation created below
+        Tool::Invocation invocation;
+        invocation.auxiliaryFiles().push_back(*moduleMapAuxiliaryFile);
+        phaseContext->toolContext().invocations().push_back(invocation);
+
+        /* Copy in the module map as part of the build. */
+        if (Tool::CopyResolver const *copyResolver = phaseContext->copyResolver(phaseEnvironment)) {
+            copyResolver->resolve(&phaseContext->toolContext(), environment, { intermediateModuleMapPath }, finalModuleMapRoot, "Ditto");
+        } else {
+            fprintf(stderr, "warning: failed to get copy tool for module map\n");
+        }
+    }
+
+    /*
+     * Handle the private module map. This module map can only be specified; it is never auto-generated.
+     */
+    std::string inputPrivateModuleMapPath = environment.resolve("MODULEMAP_PRIVATE_FILE");
+    std::string intermediatePrivateModuleMapPath = environment.resolve("TARGET_TEMP_DIR") + "/" + "module.private.modulemap";
+    if (!inputPrivateModuleMapPath.empty()) {
+        /*
+         * Copy in the private module map.
+         */
+        auto privateModuleMapAuxiliaryFile = Tool::Invocation::AuxiliaryFile(
+            intermediateModuleMapPath,
+            FSUtil::ResolveRelativePath(inputPrivateModuleMapPath, workingDirectory),
+            false);
+
+        /* Define source module map. */
+        // TODO: it would be nicer to attach this to the copy invocation created below
+        Tool::Invocation invocation;
+        invocation.auxiliaryFiles().push_back(privateModuleMapAuxiliaryFile);
+        phaseContext->toolContext().invocations().push_back(invocation);
+
+        /* Copy in the module map as part of the build. */
+        if (Tool::CopyResolver const *copyResolver = phaseContext->copyResolver(phaseEnvironment)) {
+            copyResolver->resolve(&phaseContext->toolContext(), environment, { intermediatePrivateModuleMapPath }, finalModuleMapRoot, "Ditto");
+        } else {
+            fprintf(stderr, "warning: failed to get copy tool for module map\n");
+        }
+    }
+
+    return true;
```

Comment:
In these functions, the return value is more like "success" than "did it do anything". So in those cases it would still return true since nothing went *wrong*, it just didn't do anything. It might make sense to return `false` if it can't find the copy tool, but that's a wider refactor since I haven't figured out the best way to report missing tools in general yet.